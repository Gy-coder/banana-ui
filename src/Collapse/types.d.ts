export declare interface CollapseItemProps {
  title: string;
  name: string;
  selected: Array<string>;
  change: (name: string) => void;
}
