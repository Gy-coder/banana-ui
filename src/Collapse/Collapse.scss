$grey: #ddd;
$border-radius: 4px;
.g-collapse {
  border: 1px solid $grey;
  border-radius: $border-radius;

  > .g-collapse-item {
    > .g-collapse-item-title {
      border: 1px solid $grey;
      margin: -1px;
      min-height: 32px;
      display: flex;
      align-items: center;
      padding: 0 8px;
      cursor: pointer;
    }
    &:first-child {
      > .g-collapse-item-title {
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
      }
    }
    &:last-child {
      > .g-collapse-item-title:last-child {
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
      }
    }
    > .g-collapse-item-content {
      padding: 8px;
    }
  }
}
