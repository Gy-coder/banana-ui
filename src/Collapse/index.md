---
title: Collapse
nav:
  title: Collapse
  path: /components
---

# Collapse

<code src="./demos/basic.tsx" />
<code src="./demos/single.tsx" />
