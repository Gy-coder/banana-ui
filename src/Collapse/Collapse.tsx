import React from 'react';
import './Collapse.scss';
import { CollapseItemProps } from '@/Collapse/types';

interface Props {
  selected: Array<string>;
  onChange: (newSelected: Array<string>) => void;
  multiple?: boolean;
}

const Collapse: React.FC<Props> = (props) => {
  const { selected, children, onChange, multiple = true } = props;
  const handleChange = (name: string) => {
    let selectedCopy = JSON.parse(JSON.stringify(selected));
    if (multiple === false) {
      if (selectedCopy.indexOf(name) >= 0) {
        selectedCopy = [];
      } else {
        selectedCopy = [name];
      }
    } else {
      if (selectedCopy.indexOf(name) >= 0) {
        let i = selectedCopy.indexOf(name);
        selectedCopy.splice(i, 1);
      } else {
        selectedCopy.push(name);
      }
    }
    onChange(selectedCopy);
  };
  const render = () => {
    return React.Children.map(children, (child) => {
      const childElement =
        child as React.FunctionComponentElement<CollapseItemProps>;
      return React.cloneElement(childElement, {
        selected,
        title: childElement.props.title,
        name: childElement.props.name,
        change: (name: string) => handleChange(name),
      });
    });
  };
  return <div className="g-collapse">{render()}</div>;
};

export default Collapse;
