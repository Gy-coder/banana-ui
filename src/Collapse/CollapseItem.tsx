import React, { useEffect, useState } from 'react';
import { CollapseItemProps } from '@/Collapse/types';
import classnames from 'classnames';

const CollapseItem: React.FC<CollapseItemProps> = (props) => {
  const { title, children, selected, change } = props;
  const [open, setOpen] = useState<boolean>(false);
  const handleClick = () => {
    change(props.name);
  };
  useEffect(() => {
    if (selected.includes(props.name)) setOpen(true);
    else setOpen(false);
  }, [selected]);
  return (
    <div className="g-collapse-item">
      <div className="g-collapse-item-title" onClick={handleClick}>
        {title}
      </div>
      {open && (
        <div className={classnames('g-collapse-item-content', { open })}>
          {children}
        </div>
      )}
    </div>
  );
};

export default CollapseItem;
