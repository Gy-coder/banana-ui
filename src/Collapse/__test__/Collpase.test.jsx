import {
  act,
  render,
  cleanup,
  screen,
  fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import React, { useState } from 'react';
import Collapse from '../Collapse';
import CollapseItem from '../CollapseItem';

afterEach(cleanup);

describe('测试Collapse组件', () => {
  it('renders correctly', () => {
    const Temp = () => {
      const [selected, setSelected] = useState([]);
      const onChange = (newSelected) => setSelected(newSelected);
      return (
        <Collapse selected={selected} onChange={onChange}>
          <CollapseItem title="标题一" name="1">
            内容1
          </CollapseItem>
          <CollapseItem title="标题二" name="2">
            内容2
          </CollapseItem>
          <CollapseItem title="标题三" name="3">
            内容3
          </CollapseItem>
        </Collapse>
      );
    };
    const { asFragment } = render(<Temp />);
    expect(document.querySelector('.g-collapse')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it('可以点击某个title，让它的content出现', () => {
    const Temp = () => {
      const [selected, setSelected] = useState([]);
      const onChange = (newSelected) => setSelected(newSelected);
      return (
        <Collapse selected={selected} onChange={onChange}>
          <CollapseItem title="标题一" name="1">
            内容1
          </CollapseItem>
          <CollapseItem title="标题二" name="2">
            内容2
          </CollapseItem>
          <CollapseItem title="标题三" name="3">
            内容3
          </CollapseItem>
        </Collapse>
      );
    };
    const { asFragment } = render(<Temp />);
    fireEvent.click(screen.getByText('标题一'));
    expect(screen.getByText('内容1')).toBeVisible();
    fireEvent.click(screen.getByText('标题二'));
    expect(screen.getByText('内容1')).toBeVisible();
    expect(screen.getByText('内容2')).toBeVisible();
  });
  it('带有singled', () => {
    const Temp = () => {
      const [selected, setSelected] = useState([]);
      const onChange = (newSelected) => setSelected(newSelected);
      return (
        <Collapse selected={selected} onChange={onChange} multiple={false}>
          <CollapseItem title="标题一" name="1">
            内容1
          </CollapseItem>
          <CollapseItem title="标题二" name="2">
            内容2
          </CollapseItem>
          <CollapseItem title="标题三" name="3">
            内容3
          </CollapseItem>
        </Collapse>
      );
    };
    const { asFragment } = render(<Temp />);
    fireEvent.click(screen.getByText('标题一'));
    expect(screen.getByText('内容1')).toBeVisible();
    fireEvent.click(screen.getByText('标题二'));
    expect(screen.queryByText('内容1')).not.toBeInTheDocument();
    expect(screen.getByText('内容2')).toBeVisible();
  });
});
