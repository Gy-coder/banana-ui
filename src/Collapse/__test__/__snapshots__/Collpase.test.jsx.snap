// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Collapse组件 renders correctly 1`] = `
<DocumentFragment>
  <div
    class="g-collapse"
  >
    <div
      class="g-collapse-item"
    >
      <div
        class="g-collapse-item-title"
      >
        标题一
      </div>
    </div>
    <div
      class="g-collapse-item"
    >
      <div
        class="g-collapse-item-title"
      >
        标题二
      </div>
    </div>
    <div
      class="g-collapse-item"
    >
      <div
        class="g-collapse-item-title"
      >
        标题三
      </div>
    </div>
  </div>
</DocumentFragment>
`;
